ENH: Move `loadtxt` to C for much better speed

seberg opened the pull request:
This PR is based on @WarrenWeckesser `npreadtext` which was subsequently heavily modified [here](https://github.com/BIDS-numpy/npreadtext)  (and somewhat reduced in scope).

While there is more to it, the main point is to improve the speed for parsing CSV (like) files, the included io benchmark change the following way:

<details> <summary><b>Benchmarking results</b>  </summary>

(sorted so that the `ratio` is a factor, 17 means 17 times faster)
EDIT: Last table update Jan 14 (no meaningful changes expected anymore) 
```
       before           after         ratio
     [1e6b72b4]       [e2d35064]
     <add-npreadtext>       <main>    
+     9.58±0.07ms          178±2ms    18.64  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('int64', 100000)
+         949±5μs       17.5±0.1ms    18.42  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('int64', 10000)
+      8.30±0.1ms        142±0.1ms    17.16  bench_io.LoadtxtCSVStructured.time_loadtxt_csv_struct_dtype
+      9.11±0.1ms        131±0.2ms    14.40  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('int32', 100000)
+        918±10μs       13.1±0.1ms    14.24  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('int32', 10000)
+     13.4±0.08μs          188±1μs    14.05  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('int64', 100)
+     13.1±0.05μs          143±1μs    10.91  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('int32', 100)
+      15.6±0.4ms          135±1ms     8.67  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('complex128', 100000)
+      14.6±0.1ms          127±1ms     8.67  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float32', 100000)
+      14.5±0.1ms          125±1ms     8.58  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(100000)
+     1.45±0.01ms      12.3±0.09ms     8.49  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(10000)
+     15.1±0.07ms        126±0.7ms     8.40  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float64', 100000)
+     1.48±0.02ms      12.4±0.03ms     8.39  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float32', 10000)
+     1.60±0.01ms       13.4±0.1ms     8.38  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('complex128', 10000)
+        1.52±0ms      12.5±0.05ms     8.23  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float64', 10000)
+     19.1±0.03μs        146±0.2μs     7.63  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('complex128', 100)
+      18.4±0.2μs        135±0.9μs     7.33  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(100)
+     18.7±0.08μs        137±0.5μs     7.30  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float32', 100)
+     18.9±0.06μs          137±2μs     7.21  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float64', 100)
+     1.55±0.01ms      10.3±0.05ms     6.66  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('object', 10000)
+     16.0±0.08ms          105±1ms     6.60  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('object', 100000)
+        916±10μs      5.98±0.02ms     6.52  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv([1, 3])
+     1.15±0.02ms      7.36±0.01ms     6.40  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv([1, 3, 5, 7])
+      18.2±0.3μs          110±1μs     6.03  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('object', 100)
+        798±10μs      4.80±0.01ms     6.02  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv(2)
+     5.07±0.01μs       29.7±0.7μs     5.86  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('int64', 10)
+     7.36±0.08ms       39.7±0.2ms     5.39  bench_io.LoadtxtCSVDateTime.time_loadtxt_csv_datetime(20000)
+      77.1±0.5μs          413±3μs     5.35  bench_io.LoadtxtCSVDateTime.time_loadtxt_csv_datetime(200)
+         737±7μs      3.92±0.01ms     5.32  bench_io.LoadtxtCSVDateTime.time_loadtxt_csv_datetime(2000)
+     5.06±0.02μs       25.1±0.5μs     4.95  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('int32', 10)
+     12.3±0.02μs       58.3±0.5μs     4.74  bench_io.LoadtxtCSVDateTime.time_loadtxt_csv_datetime(20)
+     5.67±0.03μs       26.3±0.4μs     4.64  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('complex128', 10)
+         220±4μs          989±4μs     4.49  bench_io.LoadtxtReadUint64Integers.time_read_uint64_neg_values(1000)
+         221±5μs          987±8μs     4.46  bench_io.LoadtxtReadUint64Integers.time_read_uint64(1000)
+         123±3μs          549±2μs     4.46  bench_io.LoadtxtReadUint64Integers.time_read_uint64_neg_values(550)
+         124±1μs          548±2μs     4.41  bench_io.LoadtxtReadUint64Integers.time_read_uint64(550)
+     2.22±0.03ms      9.76±0.04ms     4.41  bench_io.LoadtxtReadUint64Integers.time_read_uint64_neg_values(10000)
+     2.22±0.03ms      9.69±0.04ms     4.37  bench_io.LoadtxtReadUint64Integers.time_read_uint64(10000)
+     5.62±0.01μs       24.2±0.8μs     4.31  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float32', 10)
+     5.64±0.01μs       24.0±0.1μs     4.26  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('float64', 10)
+     5.44±0.03μs       22.8±0.3μs     4.20  bench_io.LoadtxtCSVComments.time_comment_loadtxt_csv(10)
+     5.90±0.02μs       21.8±0.6μs     3.69  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('object', 10)
+     4.22±0.01ms      12.7±0.08ms     3.02  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('str', 10000)
+      46.6±0.3μs        137±0.7μs     2.94  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('str', 100)
+      42.6±0.1ms        122±0.7ms     2.87  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('str', 100000)
+      73.1±0.7ms        202±0.4ms     2.76  bench_io.LoadtxtCSVSkipRows.time_skiprows_csv(0)
+      72.6±0.7ms        200±0.7ms     2.76  bench_io.LoadtxtCSVSkipRows.time_skiprows_csv(500)
+      66.1±0.9ms        182±0.6ms     2.76  bench_io.LoadtxtCSVSkipRows.time_skiprows_csv(10000)
+      10.2±0.2μs       24.3±0.1μs     2.39  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('str', 10)
```

</details>

**Marking as draft, but I do not expect major changes, so this can get a first round of review, I think.**

There is a small list of TODOs:
* [x] The current additional tests from npreadtext are missing here (they should be rewritten in terms of loadtxt probably).
* [x] Additional tests should be added (see also npreadtext issue tracker;  one reason for the PR, it makes checking code-coverage more convenient.)
* [x] Smaller changes ~(e.g. I think I want to remove passing `usecols` as a numpy array)~ Should be mostly done. ~Moving usecols could be done (and would be nice), but we don't have quite the right stuff available so added the correct sanity checks instead.  I agree this is a wart (a private one), so if anyone insists I can look at it.~ Moved usecols as well, nothing is quite awesome, but I think it is _better_.

The [issues from `nreadtext`](https://github.com/BIDS-numpy/npreadtext/issues) still apply, and we should review some of these or make a call on them.  The issue tracker there may also be a good way to open other smaller things to fix-up (to reduce noise here).

---

As a general overview, the parser has a tokenizer, that works with unicode, but supports the Python modes UCS4, UCS2, and UCS1.  It parses everything into smaller UCS4 (unicode) strings, this copies the data – in theory not always necessary, but convenient – although future optimization could be done.
After this step, we have dtype (especially for integers) specific converter functions which directly convert the UCS4 string into the final result.  If necessary, we fall back to `PyArray_Pack`.
(Since we do not discover the result `dtype`, strings have to offload some work to Python currently.)

In general, we read either line-by-line from Python iterables, or in chunks if reading from files.

The new features are currently fairly limited (e.g. I removed the custom float parser to support using `d` instead of `e`), and is not fully CVS parser featured (but this can easily be achieved now!).
One particular important changes is the current decision to remove hex float parsing and limit integer parsing to integers (we used to fall back to floats).  The custom parsers also mean that we do not support numbers with an underscore `100_000` right now.
Our current idea was that we should improve the error message a bit more, but generally advise users to use `converters=int` in this case.

Please check the `npreadtext` issue tracker for issues/changes: https://github.com/BIDS-numpy/npreadtext/issues

EDIT: One big additional feature ~(which I have not made public here, but is as easy as adding the kwarg)~, is that the parser does support quoting (with excel double-quote escaping).

seberg committed:
ENH: Move npreadtext into NumPy for faster text reading

This replaces `np.loadtxt` with the new textreader.

The file has a few minor cleanups compared to the npreadtext version.

npreadtext was started by Warren Weckesser for inclusion in NumPy
and then very heavily modified by me (Sebastian Berg) to improve it and
slim it down slightly.

Some parts of this code is inspired or even taken from the pandas parser
(mainly the integer parsers are fairly verbatim still).

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>

seberg committed:
Fixup size_t's to (mostly?) use npy_intp to silence compiler warnings

seberg committed:
MAINT: Remove float-to-int integer parsing fallback

seberg committed:
ENH: Allow a single converter to be used for all columns

This is always used if it is a callable.

seberg committed:
TST: Fixup current loadtxt tests for changes

seberg committed:
STY: Fix some style issues (mainly long lines)

Note that one of the long lines is a link that cannot be split
reasonably.

seberg committed:
MAINT: Make text reader custom mem handler compatible

seberg commented:
(The PyPy test failure should be the same as in the `empty`/intp sequence parsing PR: PyPy had a bug with formatting `%(100)R`, it is fixed, and for now we should just ignore it.)

mattip commented:
I admit I am not a big fan of converting working python code into 90% working C code by hand. It seems like the wrong approach for 2021.

I would be interested to see what happens if we use cython to convert loadtxt (and its functions) to C. Does it achieve a similar speedup?

In any case, can we leave this as an external package for now, especially given the open issues and design decisions. Perhaps this PR could target nreadtext instead of NumPy?

charris commented:
Fast reading of data in huge text files seems to be important, at least comparisons are popular :) Julia boasts it is faster than Pandas, Dask touts the advantages of parallelization, so on and so forth. I don't deal with huge text files, but use of low level languages to speed up common use cases is pretty much why NumPy and SciPy exist, so I am not opposed in principle. The only problem I see is trading off the complexity/maintainability  of a low level language vs Python. But I also note that NumPy is becoming more complex with SIMD and C++ templates, it is perhaps in the nature of things that performance drives development.

bashtage commented:
3k lines for a 50% improvement over pandas when using maximum float precision seems like a less-than-ideal ratio (and slower if one is happy with the default).

```
>>> c = pd.read_csv("test.csv", delimiter=",")
96.2 ms ± 1.45 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
>>> d = pd.read_csv("test.csv", delimiter=",", float_precision="round_trip")
322 ms ± 7.27 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> e = _loadtxt("test.csv", delimiter=",")
215 ms ± 2.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

bashtage commented:
> I would be interested to see what happens if we use cython to convert loadtxt (and its functions) to C. Does it achieve a similar speedup?

IME one big advantage of Cython over C -- especially if the Cython is pythonic and uses cython cimports rather than C function calls -- is the breadth of the base that is comfortable making PRs.

bashtage commented:
Looking at [this post](https://juliacomputing.com/blog/2020/06/fast-csv/) it seems to me that unless one wants to do a multithreaded reader then the best case gains are pretty limited (if one wants to compare to Julia as a reference point).

seberg commented:
> 3k lines for a 50% improvement over pandas when using maximum float precision seems like a less-than-ideal ratio (and slower if one is happy with the default).

Yes, IIRC, compared to pandas this has basically parity or up to 50% faster.  But, I think you will find the C code to be much more maintainable and much easier to extend.  Some notes:
* To get feature-parity with loadtxt, the pandas version must use its Python backend which is _also_ extremely complex (and very slow of course).  This is for things like:
  * Reading any file that isn't a raw utf-8 file.
  * Use any control character that is not ASCII.  E.g. splitting by whitespace `delimiter=None` doesn't work, pandas only supports ascii whitespace with any decent speed (even with `delimiter=','` non-ascii whitespace will not get stripped when other whitespace does get stripped).
  On the contrary, here the reading itself is still largely using Python API to keep it simple and flexible (with a line-by-line vs. chunked mode adding a bit of complexity for ~30% speedup, I admit).  Meaning that all of those cases are covered by the same code.
* I am very sure that if you will have much more problems extending the pandas code.  I wrote this in a way that it should be straight forward to e.g. allow users to supply parsers written in C directly, or allow swapping out our parsers (if you want to have that fast float parser). I like the idea to make it so that users can supply `converters=faster_float_parser_written_in_C`.  (yes, we would have to nail down public API for that, but the general design is there.)
* The pandas code is much less maintainable also IMO.  It has as many lines of code, but is not nearly as well structured.

Is Cython better?  I guess opinions will differ on that.  It is probably more accessible to contributions, but honestly, I am not sure I like Cython for code that is "heavy C".  It is great for the interface code, but the interface code isn't the big chunk here.

charris commented:
> 3k lines for a 50% improvement over pandas when using maximum float precision seems like a less-than-ideal ratio

Pandas has the reputation of being the best/fastest Python text reader, and a long term goal has been to make NumPy as good as Pandas. Reaching that goal probably requires doing most of the work in C. It would be interesting to see just what makes the difference when the code is C rather than Python -- reading, parsing, writing, etc. -- but it does seem clear that all the fastest readers are written in low level languages for the common case. Both R and Pandas do that, and Julia is probably in the low level language camp.

mattip commented:
What are the benchmarks for the code? Is there something that exercises the non-float dtypes?

rgommers commented:
My 2 euro cents:
- We have a ton of open `loadtxt`/`genfromtxt` issues, and this has been a mess for years. That plus the slowness justifies a rewrite. 
- I wouldn't have had a very clear preference between adopting the Pandas reader (well-tested by now) or rewriting from scratch (cleaner, bit faster code) without spending quite a bit of time looking into the problem. I'd lean towards the Pandas one mostly to avoid spending a lot of developer time. That said, now that the work is mostly done I'm happy to trust @seberg & co that this PR is the right choice by now.
- It looks like a low-stakes decision to merge this PR once it's complete. It's then faster and better structured than the Pandas reader, and benchmarks look great. So if the tests pass (and ideally it fixes a bunch of bugs as well), then +1 for merging. In the (quite unlikely) case it turns out to be horribly buggy, revert it in `1.23.1` and try again or reconsider.

> Is Cython better? I guess opinions will differ on that. It is probably more accessible to contributions, but honestly, I am not sure I like Cython for code that is "heavy C". It is great for the interface code, but the interface code isn't the big chunk here.

I tend to agree with this point of view. In addition, build system wise C is _much_ nicer.

tylerjereddy commented on numpy/core/src/multiarray/textreading/tokenize.c.src:
```diff
@@ -0,0 +1,449 @@
+ * maximum number of fields when known.
+ *
+ * Unlike other tokenizers, this one tries to work in chunks and copies
+ * data to words only when it it has to.  The hope is that this makes multiple
```
double `it`

tylerjereddy commented on numpy/lib/npyio.py:
```diff
@@ -721,110 +723,329 @@ def _savez(file, args, kwds, compress, allow_pickle=True, pickle_kwargs=None):
-            start += length
-        return tuple(ret)
+        # No preprocessing necessary
+        assert comments is None
```
Here and a few lines up---do we really want a Python `assert` in the production code to only be used in lower-optimization builds? Or would it make sense to add simple errors/error messages for these code paths if genuinely invalid?

tylerjereddy reviewed (commented):
The test suite changes look fairly modest, and include some improved error messages it seems.

seberg commented on numpy/core/src/multiarray/textreading/conversions.c:
```diff
@@ -0,0 +1,375 @@
+}
+
+/*
+ *  `item` must be the nul-terminated string that is to be
```
yeah, `\0` (as ascii or unicode).  Seems I forgot to cleanup the comment(s).  It can assume that the end of the string is a `\0`, but the string may include `\0` charcters.

seberg replied:
Just deleted this comment for now, doesn't really seem useful, all of the `to_...` functions are converters...

seberg commented:
> The test suite changes look fairly modest, and include some improved error messages it seems.

Yeah, the main changes I am aware of are:
* more restrictive parsing: no floats when storing integers, no underscore `100_000` support (can be gotten back using `converters=float` although a bit annoying if you try to use both versions, since the old one requires listing all columns.
* There is some strange stuff about counting lines vs. counting rows: gh-19785 (this can be fixed in any way pretty much, but needs a decision).  Also if you pass in rows that contain newlines, things get a bit weird.  I did not care much, because it seems super strange, but it is a change

The other things should really be improvements, like the better errors (although sometimes changed) – could still add error chaining there, especially for user provided converters.

In any case, thanks for taking a look!  I will be off mostly for the next 1.5 weeks, so probably no updates :).

seberg committed:
MAINT: Address Tylers review comments

(Mainly revising the doc strings)

seberg commented on numpy/core/src/multiarray/textreading/field_types.h:
```diff
@@ -0,0 +1,49 @@
+/*
+ * The original code had some error details, but I assume that we don't need
+ * it.  Printing the string from which we tried to modify it should be fine.
+ * This should potentially be public NumPy API, although it is tricky, NumPy
```
Pretty much fully revised this doc-string, thanks.

rossbar committed:
Port over tests from npreadtext test suite

- Add test for parsing scientific notation.
- Add multiple-char comment test.
- Port over tests for structured dtypes.
- Add tests for exceptions on skiprows/max_rows.
- port over ndmin tests.
- Make structured data reusable, add unpack tests.
- Port over delimiter tests.
- Port over maxrows test w/ various dtypes.
- Port over test of exception msg on parse failure.
- Port over test for converters w/neg indices.
- Port over usecols tests
- Port over unicode tests.
- Port over more converter tests.
- Port over test for large rows.
- Port over test for string-len discovery.
- Port over float conversion accuracy test.
- Port over bool test.
- Add test for implicit float->int conversion.
- Port over complex parsing tests.
- Port over tests for reading from generator.
- Port over object cleanup test.
- Port over bytes incompat test.
- Port over converters tests.

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>
Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

seberg committed:
BUG: Fix skiprows handling and simplify lineskipping logic

Somewhere, the skiprow handling got broken, this fixes it.  It also
simplifies the "forward to end of line" logic slightly (and hopefully
correctly).  This logic is used when a comment is reached or for
skipping lines before any actual data is read.

seberg committed:
ENH: Raise an error for (most) stray newline characters

This makes it strict that newline characters _within_ a single
line (which is only possible if the user passes in a manual iterator
of strings), is considered weird and rejected.

An example is: `np.loadtxt(['1\n1', "2 2"], dtype=np.int64)`

seberg committed:
ENH: Reject empty string as control character

`None` is forced instead in all cases (mainly applies to comments).
This is not really a change in behaviour: It was always utterly
broken.

The one weird thing about it is that `delimiter=None` means
"any whitespace", while `quote=None` and `comments=None` means
that no quote/comment character exists at all.

seberg committed:
TST: Small fixups for tests to make sure they pass again

seberg committed:
TST: Fix test to align with stricter integer parsing

xor2k commented:
I can't resist to comment here, so sorry in advance ;)

Long time ago I have written my own CSV parser. It had support for quotes, I think that should be pretty much standard. CSV parsers are just inherently complex and need to handle legacy, since this is what CSV is about to a large amount ("I want to read that old file from this legacy system XYZ here"). I totally agree with @seberg that it is about the structure of the code rather than the language it is written in (C or Python) to keep this complexity as low as possible and a CSV parser maintainable. Maintainability and reliability should be more important than speed anyway, since there is not so much speed to be gained, as even GPUs can only accelerate by a factor of 2-3x, compare

https://medium.com/rapids-ai/reading-larger-than-memory-csvs-with-rapids-and-dask-e6e27dfa6c0f

Having CSV in a workflow is frustrating. It is always slow and when dealing with floats (so probably most of the time), it is a reason for precision loss. In my workflows, I typically convert CSV to HDF or .npy first and create intermediate, binary cache files if needed to have short cycle times ("save .py, automated run (e.g. with entr), see results" should take only a few seconds). However, one nice thing about CSV is that it can be easily appended to, just by writing to the end of the file and at this point, I'd like to advertise my pull request a little more:

https://github.com/numpy/numpy/pull/20321

This would make .npy appendable in the same way CSV can be appended to: just write some more data to the end of the file, no need to update headers.

xor2k commented:
And I'd also like to add that most CSV parsers I've encountered (C, C++, JavaScript, Java) are simply broken and unmaintained. The one from Pandas is indeed the best I've seen so far, probably because of its large user base.

rossbar committed:
Rename quotechar param and update docstring.

rossbar committed:
TST: Add tests for quote character support.

rossbar committed:
Add test to check quoting support disabled by default.

rossbar committed:
Add tests for quote+multichar comments.

Also correct exception message.

rossbar committed:
TST: structured dtype w/ quotes.

rossbar committed:
Add tests for empty quotes and escaped quotechars.

rossbar committed:
rm incorrect comment.

seberg committed:
BUG: Add missing quote copy

The code that copied the quote character in when a double-quote
occurred went AWOL.  Add a copy path back.

seberg committed:
DOC: Add release notes for loadtxt changes

These are pretty verbose right now, so we may wish to move some of
it into the docs or similar.
I think they cover most/all important changes, although some very
corner cases changes are currently ignored (e.g. we don't support
embedded newlines in rows...)

mattip commented:
> We have a ton of open loadtxt/genfromtxt issues

It might be nice to see which of the 60 open issues with [loadtxt or genfromtxt](https://github.com/numpy/numpy/issues?q=is%3Aissue+is%3Aopen+genfromtxt+OR+loadtxt) are relevant

seberg commented:
Frankly, I don't think it closes a lot.  It closes some performance (speed/memory) relevant ones (although some of them mention "genfromtxt" as well):
* gh-7480
* gh-8523 (for loadtxt)

It fixes/changes gh-19785 (as documented and we wanted to clean this part up slightly more), but I can also undo that. It also implements gh-9291, but it is just a small carrot if anything.  A major thing is quoting support as e.g. requested in gh-2211 (but not for genfromtxt for now obviously).
I think there are a couple of others that may just not be listed anywhere, like:
* Using `float` for `int` conversions is clearly incorrect for some values.
* Better error messages

So does it fix a huge amount of issues? No, not really.  It addresses the few big ones that meant that everyone is told not to use loadtxt: performance/memory use and CSV features like quoting – full features can now be achieved, such as escaping.
For most issues, moving it to C doesn't really change all that much about how hard/easy it would be to address them.

Warren had added support for things like `decimal=","`, I ripped it out because that chunk relied on vendoring 3000 lines of float parsing code and I prefer to start minimal.  But that can be re-added also without vendoring, by adding it to a translation step (which really should not make a big difference on the scale of float parsing).

seberg committed:
MAINT: Replace last uses of raw malloc with PyMem_MALLOC

rossbar committed:
Add UserWarning when reading no data.

seberg committed:
MAINT: Fixup include guards and use NPY_NO_EXPORT (or static) throughout

seberg committed:
MAINT: Add sanity check to ensure usecols is correct.

rossbar committed:
Add warning on empty file + tests.

seberg committed:
BUG: Fix complex parser and add tests for whitespace and failure paths

seberg committed:
BUG,TST: Add test for huge-float buffer path and ensure error return

If a memory error happens, we should at least not crash the interpreter

seberg committed:
TST: Add test to cover copyswap (byte-swap and unaligned)

seberg committed:
DOC: See if adding a newline fixes release note rendering

seberg committed:
BUG: Fix some issues found by a valgrind run

rossbar committed:
Update and add converters examples.

seberg committed:
BUG: Fix growing when NPY_RELAXED_STRIDES_DEBUG=1 is used

rossbar committed:
Add quotechar to examples.

seberg committed:
MAINT: Move usecol handling to C and support more than integer cols

Of course to actually use that many columns you need A LOT of memory
right now.  Each field stores at least a UCS4 NUL character, but
the field is padded enough to require 16 bytes.
We always parse a full row, so that requires 20 bytes per field...
(i.e. 32 GiB RAM is not enough to test this :)).

seberg committed:
BUG: Make sure num-fields is intp/ssize_t compatible

In theory (if homogeneous) we actually should support more than
2**31 columns.  This should fix that.

Also cap overallocation scheme, so that we don't waste quite so
much memory in these extreme cases

seberg committed:
BUG: Ensure current num fields holds enough space for ultra-wide columns

A bit tricky, the code used to assume we are limited to <32bits due to the
dtype limitations.  But we are not for homogeneous arrays, so lets try
to get it to work.
(Right now it seems to fail, although even if it succeeds it will need
a huge amount of memory)

seberg committed:
ENH: Give warning for empty-lines not counting towards max-rows

This also slightly cleans up the empty-line handling: previously
we sometimes just had some extra empty-lines that just effectively
never mattered

seberg committed:
MAINT: Small cleanup, use FINLINE for int parsers

It seems the boolean converter using the integer parser caused it
to not always be inlined as desired, so just use FINLINE.

seberg committed:
MAINT,TST,BUG: Simplify streamer init, fix issues, and add tests

seberg committed:
TST,BUG: Additional bad-file-like test, add missing error path free

And remove one silly leftover struct member that was unused

seberg committed:
TST,MAINT: New tests, byteswap cleanups and fixed assert

seberg committed:
TST: Improve test coverage, replace impossible error with assert

seberg committed:
MAINT: Remove unused/unnecessary allow-embedded-newlines option

No other parser has this option, so I think we really do not need it.

seberg committed:
TST: Add test for hard/impossible to reach universal-newline support paths

seberg committed:
MAINT: Use skiplines rather than skiprows internally throughout

Skiplines is just the more clear names since "rows" make a lot of
sense for output rows (which implies that a line is not empty for
example)

seberg commented:
<details> <summary> TL;DR: Even trying very hard to move to Python's `csv` seems not that worthwile and should leave us also with a 2× speed disadvantage. </summary>

Btw., I had a sanity check and using python's `csv` would be at least 2x slower. There are two (conservative) estimates:
* Use `dtype=object, encoding="utf-8"` to use our parser to generate an array of unicode objects
* Use `csv.reader` exhausting it with `[None for x in (csv.reader(...))]` (so we do not pile up unicode objects like crazy, which is much slower as it has to allocate new objects all the time, rather than using the unicode free-list)

Both is >2× slower (and surprisingly very similar in performance – it looks like the Python's reader may have an almost ~2× performance improvement possible by using this tokenizer).
Even optimizing Python's parser, I don't really see it likely that we can remove a 2× (or more) disadvantage, and even that can only remove the tokenizer and not the custom conversion side of it.
</details>

For testing: I believe my latest push gives us 100% of possible C code coverage, but a few odder behaviour tests may still be desirable/missing.
This should be ready very soon (I assume next week), aside possible cleanups of tests or code cleanup as part of review.

For the `max_rows` vs. `max_lines` behaviour, I have added a `UserWarning` (which may still be modified slightly).  There are two further deprecations we would like to do, but I now prefer pushing off:
* `encoding="bytes"` is really ripe for deprecation/removal.  Lets force users to deal with it finally by deprecating its use?
* It would be nice to rename `skiprows` to `skiplines` because that is closer to the truth.

seberg committed:
MAINT: Very minor style cleanups (mostly)

I.e. no use of goto where it is not necessary.

seberg committed:
MAINT: Only allocate converters if necessary

This actually *slows* things down very mildly, but it just seems
a bit cleaner to me, so lets do it anyway...

seberg committed:
TST: Move most new loadtxt tests to its own file

This also adds two basic new tests around files/strings containing
the \0 character (prooving that we handle that gracefully).

Also adds tests for:
* the `_` thousands delimiter (should fail, but doesn't for float128
  right now)
* Failure modes when the number of rows changes (negative specifically)

Many of these tests came originally from Warren Weckesser and others
were added by Ross Barnowsky:

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>
Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

seberg committed:
TST,STY: Add small additional tests for converters/usecols

Also fix style a bit to silence linter (hopefully), removes some
black style, but I am not too opinionated about that :)

seberg committed:
DOC: Remove outdated loadtxt TODOs from code

seberg committed:
BUG: Fix loadtxt no data warning stacklevel

seberg committed:
TST,BUG: Fortify byteswapping tests and make a small fix

I had a lingering feeling I should double check this, turns out
that feeling was right ;).

(Fixes up the tokenizer doc a bit.)

seberg committed:
ENH: Give a clear error when control characters match/are newlines

These never have a meaning (at best an implicit order meaning one
is ignored), except theoretically in the `delimiter=None` case
(or in strange cases with `\r` being a delimiter, but `\n` being
the newline when a file is opened manually but not in universal
newline mode).

It seems more useful to just generally raise an error, since all of
"features" are weird corner cases and likely surprising to users.

mattip commented on numpy/lib/tests/test_loadtxt.py:
```diff
@@ -0,0 +1,856 @@
+    with pytest.raises(TypeError, match="keys of the converters dict"):
+        np.loadtxt(StringIO("1 2\n3 4"), converters={"a": int})
+    with pytest.raises(TypeError, match="keys of the converters dict"):
+        np.loadtxt(StringIO("1 2\n3 4"), converters={"a": int}, usecols=0)
```
This test causes a segfault on PyPy because of a bug in the current PyPy version. I think the best way around that is to add (with proper linting)
```
@pytest.mark.skipif(sys.implementation.name == 'pypy' and sys.implementation.version <= (7, 3, 8), reason="PyPy bug in error formatting")
```

seberg committed:
TST: Skip unparsable field error tests on PyPy

PyPy has a small bug with error formatting, so these cause it to
crash.  Simply skip the tests on old PyPy versions for now.
(Matti fixed the issue long ago, just waiting for a new PyPy release :))

seberg committed:
TST: Use hand-picked values for byte-swapping tests

Longdouble is a source of problems here especially (mainly due to
it sometimes using double in the background, or maybe just buggy
implementations).  Together with strings that correctly parsed
do not roundtrip (if printed using less precision), things just do
not work out...  This fixes it, and is simpler/clearer anyway.

seberg committed:
TST: Catch two more errors that runs into the PyPy issue

seberg committed:
TST: Use repr in byteswapping tests

The `str` values for those weird values used for longdouble are
truncated by PyPy's complex `str` output.  Which seems fine probably
since PyPy's `repr` does the right thing and will not truncate.

rossbar committed:
Add test for datetime parametric unit discovery.

rossbar committed:
TST: Some tests for control character collisions.

Adds some tests for the behavior of control characters, e.g. comments, delimiter and quotechar, when they have the same value. At this stage, these tests are more to frame the discussion about what the behavior should be, not to test what it currently is. I personally think raising an exception is correct for most of these situations, though it's worth noting that np.loadtxt currently doesn't for most of these corner cases (and seems to randomly assign precedence to delimiter over comments or vice versa depending on the values).

rossbar committed:
Add test for unicode, parametrize for chunksize.

rossbar committed:
Add test for empty string as control characters.

Includes comments param, which is handled on the Python side.

rossbar committed:
Add test for str dtype len discovery with converters.

nrows gt chunksize.

rossbar committed:
Handle delimiter as bytes.

rossbar committed:
Linting.

rossbar commented:
The last of the small fixups that were queued up from previous reviews have been addressed, so this one is ready for review!

rossbar committed:
TST: Fix exception msg matching in tests.

seberg committed:
TST: Skip error test using on PyPy (test uses %.100R)

rossbar committed:
Add two new examples of converters to docstring examples
 - Floats with underscores
 - Floats + hex floats.

mattip commented:
something went wrong with CI

mattip reviewed (approved):
We discussed this today and it is ready for merging once CI is green. There will be some follow-on PRs but this looks good so far. Thanks @WarrenWeckesser, @rossbar and @seberg.

seberg commented:
Thanks Matti!  A typo (or two) snuck into the last docs update; fixed now. the remaining failure is due to the scipy docs (general doc CI failure currently).

> There will be some follow-on PRs but this looks good so far.

Yes, the main potential follow-ups will be looking into:
* removing that Python 2 compatibility `encoding="bytes"` shim.
* renaming `skip_rows` to `skip_lines` to make it clearer.

(Besides any potential issue fixes of course. I will probably send a brief heads-up email to the mailing list once it lands.)

mattip commented:
Thanks @seberg, @rossbar, @WarrenWeckesser
